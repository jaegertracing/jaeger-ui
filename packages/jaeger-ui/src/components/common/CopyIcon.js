// @flow

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { Icon, Tooltip } from 'antd';
import { CopyToClipboard } from 'react-copy-to-clipboard';

type PropsType = {
  className?: string,
  copyText: string,
  tooltipTitle: string,
};

type StateType = {
  hasCopied: boolean,
};

export default class CopyIcon extends React.PureComponent<PropsType, StateType> {
  static defaultProps = {
    className: null,
  };
  state = {
    hasCopied: false,
  };

  handleClick = () => {
    this.setState({
      hasCopied: true,
    });
  };

  handleTooltipVisibilityChange = (visible: boolean) => {
    if (!visible && this.state.hasCopied) {
      this.setState({
        hasCopied: false,
      });
    }
  };

  render() {
    return (
      <Tooltip
        arrowPointAtCenter
        mouseLeaveDelay={0.5}
        onVisibleChange={this.handleTooltipVisibilityChange}
        placement="left"
        title={this.state.hasCopied ? 'Copied' : this.props.tooltipTitle}
      >
        <CopyToClipboard text={this.props.copyText}>
          <Icon className={this.props.className} onClick={this.handleClick} type="copy" />
        </CopyToClipboard>
      </Tooltip>
    );
  }
}
