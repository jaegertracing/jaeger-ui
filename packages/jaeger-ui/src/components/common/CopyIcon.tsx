// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import { Button, Tooltip } from 'antd';
import { TooltipPlacement } from 'antd/es/tooltip/index';
import cx from 'classnames';
import copy from 'copy-to-clipboard';

import './CopyIcon.css';
import { IoCopyOutline } from 'react-icons/io5';

type PropsType = {
  className?: string;
  copyText: string;
  icon?: React.ReactNode;
  placement?: TooltipPlacement;
  tooltipTitle: string;
  buttonText: string;
};

type StateType = {
  hasCopied: boolean;
};

export default class CopyIcon extends React.PureComponent<PropsType, StateType> {
  static defaultProps: Partial<PropsType> = {
    className: undefined,
    icon: <IoCopyOutline />,
    placement: 'top',
  };

  state = {
    hasCopied: false,
  };

  handleClick = () => {
    this.setState({
      hasCopied: true,
    });
    copy(this.props.copyText);
  };

  handleTooltipVisibilityChange = (visible: boolean) => {
    if (!visible && this.state.hasCopied) {
      this.setState({
        hasCopied: false,
      });
    }
  };

  render() {
    return (
      <Tooltip
        arrowPointAtCenter
        mouseLeaveDelay={0}
        onOpenChange={this.handleTooltipVisibilityChange}
        placement={this.props.placement}
        title={this.state.hasCopied ? 'Copied' : this.props.tooltipTitle}
      >
        <Button
          className={cx(this.props.className, 'CopyIcon')}
          htmlType="button"
          icon={this.props.icon}
          onClick={this.handleClick}
        >
          {this.props.buttonText}
        </Button>
      </Tooltip>
    );
  }
}
