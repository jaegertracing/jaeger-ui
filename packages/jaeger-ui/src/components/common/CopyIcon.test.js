// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Icon, Tooltip } from 'antd';

import CopyIcon from './CopyIcon';

describe('<CopyIcon />', () => {
  const props = {
    className: 'classNameValue',
    copyText: 'copyTextValue',
    tooltipTitle: 'tooltipTitleValue',
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CopyIcon {...props} />);
  });

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('updates state when icon is clicked', () => {
    expect(wrapper.state().hasCopied).toBe(false);
    wrapper.find(Icon).simulate('click');
    expect(wrapper.state().hasCopied).toBe(true);
  });

  it('updates state when tooltip hides and state.hasCopied is true', () => {
    wrapper.setState({ hasCopied: true });
    wrapper.find(Tooltip).prop('onVisibleChange')(false);
    expect(wrapper.state().hasCopied).toBe(false);

    const state = wrapper.state();
    wrapper.find(Tooltip).prop('onVisibleChange')(false);
    expect(wrapper.state()).toBe(state);
  });

  it('persists state when tooltip opens', () => {
    wrapper.setState({ hasCopied: true });
    wrapper.find(Tooltip).prop('onVisibleChange')(true);
    expect(wrapper.state().hasCopied).toBe(true);
  });
});
