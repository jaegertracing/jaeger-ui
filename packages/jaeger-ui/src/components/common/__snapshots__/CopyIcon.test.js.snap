// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CopyIcon /> renders as expected 1`] = `
<Tooltip
  arrowPointAtCenter={true}
  autoAdjustOverflow={true}
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.5}
  onVisibleChange={[Function]}
  placement="left"
  prefixCls="ant-tooltip"
  title="tooltipTitleValue"
  transitionName="zoom-big-fast"
>
  <CopyToClipboard
    text="copyTextValue"
  >
    <Icon
      className="classNameValue"
      onClick={[Function]}
      type="copy"
    />
  </CopyToClipboard>
</Tooltip>
`;
